CI: Add AKS helm overrides for E2E test

vipul-21 opened the pull request:
AKS does not have the E2E integration enabled for AKS. This PR adds `aks` as a valid `CNI_INTEGRATION` attribute. The change includes the helm overrides and the test that should be skipped for AKS.

joestringer commented on test/helpers/kubectl.go:
```diff
@@ -153,6 +156,40 @@ var (
 
+	aksHelmOverrides = map[string]string{
+		"image.repository":          "mcr.microsoft.com/oss/cilium/cilium",
+		"image.tag":                 "1.12.1",
```
Could you parameterize these `repository` + `tag` references? I think it makes sense to be able to launch the testsuite against target development images like this, but given that the Cilium CNCF project doesn't have control over these Microsoft repositories, it is a bit odd to hardcode these references into the code tree. Also it would be useful to not have to update the version string here every time that we want to run tests against a different version.

In other cases, we have just defined these via environment variables and passed them in as overrides.

vipul-21 replied:
Sure, make sense. So you want the test case to have a default image and tag, but can be passed through parameters as well. Is that correct ?

joestringer replied:
Yeah that's what I had in mind. I suspect there's already some way to do this already, because the existing cloud jobs would need to pull the image from somewhere (which is typically from cilium/cilium-ci).

joestringer commented on test/helpers/kubectl.go:
```diff
@@ -153,6 +156,40 @@ var (
+		// # kubenet pod CIDR
+		"k8s.requireIPv4PodCIDR":              "false",
+		"ipv6.enabled":                        "false",
+		"ipv4NativeRoutingCIDR":               "10.241.0.0/16",
```
I'm not sure exactly how this works but if this often varies in the cluster depending on how you provisioned it, then maybe this one should be parameterized as well. This is already parameterized for GKE runs (`grep NATIVE_CIDR` in the test/ dir).

vipul-21 replied:
Regarding this a well. Make the default value as this and pass it as parameter through cmd line ? 
I saw the GKE code, they are fetching it from the ENV Variable. What should be the ideal way to pass the parameterized this ?

vipul-21 replied:
Also, how will this affect the test cases being run in the PR ? ( having a parametric value in the test cases.)

joestringer replied:
Environment variables are usually an easy way to do this, because you can then just set up your own environment with the relevant fields and it will just work. This works for either CI type environments or your local development machine.

I'm not quite sure I understand your concern around parametric values. For the IPv4 Native Routing CIDR, I would expect that if it is configured correctly, all tests pass and if it's not configured correctly then some particular forwarding path may experience problems because e.g. masquerading is misconfigured.

joestringer commented on test/helpers/kubectl.go:
```diff
@@ -1700,7 +1737,6 @@ func (kub *Kubectl) PodsHaveCiliumIdentity(namespace, podSelector string) error
 // ready, then the number of replicas is returned.
 func (kub *Kubectl) DeploymentIsReady(namespace, deployment string) (int, error) {
 	fullName := namespace + "/" + deployment
-
```
nit: These whitespace changes seem unnecessary.

vipul-21 replied:
Sure will update the PR.

joestringer commented on stable.txt:
```diff
@@ -1 +1 @@
-v1.12.1
+v1.12.1
```
I would suggest not changing this file since it will frequently conflict and change.

vipul-21 replied:
Sure will update the PR.

vipul-21 replied:
I think the version got updated to v.1.12.2. So will resolve the conflict.

joestringer reviewed (changes requested):
Thanks for the submission. Just a couple of suggestions mainly around parameterizing options that will often change, plus some other nits here and there.

joestringer commented on Documentation/contributing/testing/e2e.rst:
```diff
@@ -525,6 +525,24 @@ cluster.
+1. Setup a cluster as in :ref:`k8s_install_quick` or utilize an existing
+   cluster.
+   .. note:: You do not need to deploy Cilium in this step, as the End-To-End
+          Testing Framework handles the deployment of Cilium.
```
nit, this doesn't need to be a note.

```suggestion
   cluster. You do not need to deploy Cilium in this step, as the End-To-End
   Testing Framework handles the deployment of Cilium.
```

joestringer commented on stable.txt:
```diff
@@ -1 +1 @@
-v1.12.2
+v1.12.2
```
Please drop the changes to this file, there's no need to modify it.

joestringer reviewed (changes requested):
Overall the changes look good, couple of minor nits below. Please also drop the merge commit and instead just `git rebase` on top of the latest master branch.

joestringer commented on Documentation/contributing/testing/e2e.rst:
```diff
@@ -525,6 +525,24 @@ cluster.
+
+.. code-block:: shell-session
+  export NATIVE_CIDR="10.241.0.0/16"
+  CNI_INTEGRATION=aks K8S_VERSION=1.17 ginkgo --focus="K8s" --tags=integration_tests -- -cilium.provision=false -cilium.kubeconfig=`echo ~/.kube/config` -cilium.passCLIEnvironment=true -cilium.image="mcr.microsoft.com/oss/cilium/cilium" -cilium.tag="1.12.1" -cilium.operator-image="mcr.microsoft.com/oss/cilium/operator" -cilium.operator-suffix=""  -cilium.operator-tag="1.12.1"
```
I think that the indendation might not be right here either, typically I think we use indendation of 4 spaces for the code below a `.. code-block: ...` since sphinx/rST is opinionated about this. You can use `make render-docs` to see how the page is rendered.

vipul-21 committed:
CI: AKS E2E Integration tests

This adds the helm overrides used by the AKS cluster for the e2e tests.
Skips the ginkgo tests which are not relevant for the AKS cluster.

Signed-off-by: vipul-21 <vipul21sept@gmail.com>

joestringer reviewed (approved):
LGTM, thanks!

christarazi commented on test/helpers/utils.go:
```diff
@@ -518,6 +518,10 @@ func GKENativeRoutingCIDR() string {
 	return os.Getenv("NATIVE_CIDR")
 }
 
+func AKSNativeRoutingCIDR() string {
```
Nit: Seems like this does the same thing as `GKENativeRoutingCIDR()`, maybe consolidate these functions into `NativeRoutingCIDR()` instead?

vipul-21 replied:
I thought of doing so. I am not fully versed with the impact it might have on changing the name `GKENativeRoutingCIDR()` to `NativeRoutingCIDR()`. There is only use one case for that function. Not sure if that'll break something for GKE test cases.

Maybe want to keep GKE and AKS CIDR function segregated?

christarazi replied:
Both of them rely on the `NATIVE_CIDR`env var anyway  so there's no special logic or unexpected impact that I can see.

joestringer replied:
It's configuring the same option, so as long as it's reading the same environment variable I think this should be OK.

christarazi commented on test/k8s/updates.go:
```diff
@@ -48,6 +48,8 @@ var _ = Describe("K8sUpdates", func() {
 	)
 
 	BeforeAll(func() {
+
```
Nit: No need for extra line
